### PR TITLE
Convert ComHost tasks to use System.Text.Json instead of Newtonsoft.Json

### LINKED ISSUE #3081: migrate newtonsoft.json to system.text.json in RegFreeComManifest.cs and ClsidMap.cs
https://github.com/dotnet/sdk/blob/afae0615d5cbb16c058c5ce1fd65215d07df24d9/src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
https://github.com/dotnet/sdk/blob/afae0615d5cbb16c058c5ce1fd65215d07df24d9/src/Tasks/Microsoft.NET.Build.Tasks/RegFreeComManifest.cs

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fixes #3081 

cc: @nguerrera @wli3

### CODE REVIEW

File: eng/Versions.props [hunk 1]
```diff
@@ -15,7 +15,7 @@
     <MicrosoftBuildUtilitiesCoreVersion>15.4.8</MicrosoftBuildUtilitiesCoreVersion>
     <MicrosoftExtensionsDependencyModelVersion>2.1.0-preview2-26306-03</MicrosoftExtensionsDependencyModelVersion>
     <NETStandardLibraryNETFrameworkVersion>2.0.1-servicing-26011-01</NETStandardLibraryNETFrameworkVersion>
-    <NewtonsoftJsonVersion>9.0.1</NewtonsoftJsonVersion>
+    <MicrosoftBclJsonSourcesVersion>4.6.0-preview3.19128.7</MicrosoftBclJsonSourcesVersion>
```
Comment (wli3) [inline eng/Versions.props]: Do you need this? it should be part of runtime already

File: src/Tasks/Microsoft.NET.Build.Tasks/ArrayBufferWriter.cs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.NET.Build.Tasks
+{
+    /// <summary>
+    /// Implementation of System.Buffers.IBufferWriter backed by ArrayPool.Shared.
+    /// </summary>
+    internal class ArrayBufferWriter : IBufferWriter<byte>, IDisposable
```
Comment (wli3) [inline src/Tasks/Microsoft.NET.Build.Tasks/ArrayBufferWriter.cs]: please add a comment like `// TODO Remove once we have https://github.com/dotnet/corefx/issues/34894`

File: src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj [hunk 3]
```diff
@@ -46,6 +47,10 @@
     <PackageReference Include="Microsoft.Extensions.DependencyModel" Version="$(MicrosoftExtensionsDependencyModelVersion)" />
     <PackageReference Include="NuGet.ProjectModel" Version="$(NuGetProjectModelVersion)" />
     <PackageReference Include="NuGet.Build.Tasks.Pack" Version="$(NuGetBuildTasksPackVersion)" ExcludeAssets="All" />
+    <PackageReference Include="Microsoft.Bcl.Json.Sources" Version="$(MicrosoftBclJsonSourcesVersion)" GeneratePathProperty="true">
```
Comment (wli3) [inline src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj]: you should not need this

**Review by wli3 — CHANGES_REQUESTED** [review]
We should not need to reference the nuget package, let me check that, rest is good

File: eng/Versions.props (hunk 1 above)
Comment (wli3) [reply eng/Versions.props]: ah we need that for full framework. @dsplaisted need your opinion, should we take system.text.json package dependency? And we should set up automatic insertion

**Comment by wli3** [comment]
We need more discussion for adding system.text.json package. So i won't rush it to tonight's release 4 deadline

File: src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj (hunk 3 above)
Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj]: If we are going to use this package, we should only reference it for .NET Framework.

Also, why is this using a sources package instead of a library?

**Review by dsplaisted — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
I don't know where the package is that isn't the sources package, so I grabbed the sources package. Where is the library package located?

File: src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj (hunk 3 above)
Comment (ahsonkhan) [reply src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj]: You would need to reference it for both of the TFMs here (net472, netcoreapp2.0). The library is built-in on netcoreapp3.0.

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs [hunk 4]
```diff
@@ -189,12 +187,46 @@ private static string GetTypeGuid(MetadataReader reader, TypeDefinition type)
                 if (reader.StringComparer.Equals(attributeType.Namespace, "System.Runtime.InteropServices") && reader.StringComparer.Equals(attributeType.Name, "GuidAttribute"))
                 {
                     CustomAttributeValue<KnownType> data = attribute.DecodeValue(new TypeResolver());
-                    return (string)data.FixedArguments[0].Value;
+                    return $"{{{(string)data.FixedArguments[0].Value}}}";
```
Comment (ahsonkhan) [inline src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: I am curious, what's the reason for this change?

File: src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj (hunk 3 above)
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj]: I think we can move to 3.0.

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs (hunk 4 above)
Comment (jkoritzinsky) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: These GUIDs should be wrapped in curly brackets. It's something I noticed they weren't while I was working on this.

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs [hunk 5]
```diff
@@ -189,12 +187,46 @@ private static string GetTypeGuid(MetadataReader reader, TypeDefinition type)
...
+                json.WriteString("assembly", clsid.Value.Assembly);
+                json.WriteString("type", clsid.Value.Type);
+
+                json.WriteEndObject();
+
+                prevBytesWritten = json.BytesWritten - prevBytesWritten;
+                if (prevBytesWritten > SyncWriteThreshold)
+                {
+                    json.Flush(isFinalBlock: false);
+                    buffer.CopyTo(clsidMapFile);
```
Comment (ahsonkhan) [inline src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: I believe there is bug in the sample which will cause issues here.

After flushing, you need to reset the json writer local by re-instantiating it.

If you are keeping this `SyncWriteThreshold` pattern here (which in general should not be needed), change it to the following.

Add:
```C#

            foreach (KeyValuePair<string, ClsidEntry> clsid in map)
            {
                json.WriteStartObject(clsid.Key);

                json.WriteString("assembly", clsid.Value.Assembly);
                json.WriteString("type", clsid.Value.Type);

                json.WriteEndObject();

                if (json.BytesWritten > SyncWriteThreshold)
                {
                    json.Flush(isFinalBlock: false);
                    buffer.CopyTo(clsidMapFile);
                    json = new Utf8JsonWriter(buffer, json.GetCurrentState());
                }
            }
```

I'll fix the sample docs and call out that this pattern is mainly useful for async writes, not synchronous ones. Thanks!

Comment (ahsonkhan) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: Btw, feedback/contribution to the porting guide would be most welcome so it can be improved :)

https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/porting_guide/README.md

File: src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj [hunk 6]
```diff
@@ -46,6 +47,10 @@
...
+      <PrivateAssets>All</PrivateAssets>
+    </PackageReference>
+    <PackageReference Include="System.Memory" Version="4.5.2" />
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj]: I'm not super kean on more copy-local assemblies for the desktop build. We ran into issues with System.Memory and S.RC.Unsafe and mismatched versions in PresentationBuildTasks.dll. I don't know if it's worth adding more dependencies if we can't remove NewtonSoft.Json. And we can't because NuGet depends on it.

Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj]: I think I'd wait for our dependencies to stop using newtonsoft.json. This is going to just create more DLLs to ship and load, not fewer.

File: src/Tasks/Microsoft.NET.Build.Tasks/ArrayBufferWriter.cs (hunk 2 above)
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ArrayBufferWriter.cs]: Honestly, spreading this to every repo is getting silly. Let's wait for it to be in. I don't understand the rush to use this API. Particularly because we can't actually remove N.J until NuGet does.

**Review by nguerrera — CHANGES_REQUESTED** [review]
Thanks for doing this, but I don't think we should take this yet. I see more dlls deployed and loaded, plus a bunch of copied code that comes with TODO debt. I don't see any actual value for now. At some point, the value will come but we're not there yet.

File: src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj (hunk 6 above)
Comment (ahsonkhan) [reply src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj]: > I think we can move to 3.0.

The decision to port is up to you, but for clarity, if this moves to 3.0 and the built-in library is being used, then there won't be extra copy-local assemblies like System.Memory.

Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj]: It can't be 3.0 only. It can be 3.0 + 472. We have to run in Framework for VS msbuild

**Comment by jkoritzinsky** [comment]
I'm going to close this for now since we aren't going to merge this in for quite a bit. I'll re-open it when we're closer to taking it.

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs [hunk 7]
```diff
@@ -189,12 +187,46 @@ private static string GetTypeGuid(MetadataReader reader, TypeDefinition type)
...
-                    return (string)data.FixedArguments[0].Value;
+                    return $"{{{(string)data.FixedArguments[0].Value}}}";
                 }
             }
             throw new BuildErrorException(Strings.ClsidMapExportedTypesRequireExplicitGuid, GetTypeName(reader, type));
         }
 
+        private static void WriteClsidMap(Stream clsidMapFile, Dictionary<string, ClsidEntry> map)
+        {
+            const int SyncWriteThreshold = 1_000; // Copy to stream after writing this many bytes
```
[the remark this answers is not in the document]
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: I find it quite troubling that we're expecting every consumer to think at this level. The BCL should provide a way to write Json to a Stream, IMHO.

[the remark this answers is not in the document]
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: Where does 1 MB guideline for synchronous case come from? How would we know to choose that?

[the remark this answers is not in the document]
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: (If we were writing a big payload)

[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: The 1 MB guidelines is an implementation detail of using ArrayBufferWriter which is backed by the array pool. You can definitely continue to write up to 2 GB until OOM, but beyond 1 MB, the array pool starts to allocate. That's why I suggested, that IF you were to set a sync threshold, 1 MB is reasonable. You don't have to, though, especially if allocations aren't a concern (writing to a growable stream without a sync threshold, for instance memory stream, has the same concern).

[the remark this answers is not in the document]
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: That's fine if you need the control, but as you spread this to every consumer of NewtonSoft.Json in the stack, you're going to hit many that don't. Not being able to just write Json to a file is very odd for many of these components.

[the remark this answers is not in the document]
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: >  (writing to a growable stream without a sync threshold, for instance memory stream, has the same concern).

I would expect to be able to switch to FileStream from MemoryStream in that case, not start worrying about buffering myself.

[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: That's reasonable feedback. We made an deliberate trade-off between usability and performance. Writing to a (file)stream is definitely a common scenario, and if I can make it easier without having the lowest layer of the JSON stack (i.e. JsonWriter) regress in performance, then I would. However, since our primary focus was not high usability, but rather an emphasis on performance, the design required synchronous buffered writes (to an IBufferWriter) with periodic streamed-writes to be done by the caller (which I don't think we can change at this layer). FWIW, the serializer layer, built on top of the writer, let's you write to a stream.

[the remark this answers is not in the document]
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: It feels like there should be a thing between serialization and this layer that lets me write forward-only to a Stream. I don't get how that would regress the lower layer, which someone could still use if they really need it.

[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: We could certainly consider adding such a type in the future. Though, it would likely contain much of the same API surface as the lowest layer (overloads like WriteString, WriteNumber, etc.), but those would just end up writing directly to the stream rather than an internal buffer. The caller of the writer is actively calling the write methods today, so I don't necessarily see periodically flushing to a stream (by tracking the `BytesWritten` property) as such a significant pain-point, to justify the feature at the moment, tbh. But, I am open to the feedback, and will consider it. Thanks.

[the remark this answers is not in the document]
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: Could terre be `StreamBufferWriter : IBufferWriter`?

[the remark this answers is not in the document]
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: Maybe what I really want are the higher level things like serialization or DOM and not a stream binding for forward-only writing. It's surprising that this doesn't exist, but it's probably still lower than what we'd use if there were a full JSON stack available.

System.Reflection.Metadata (which I maintained) has similar tradeoffs. It is a low level API by design upon which higher levels can be built. I often told people that it was not the direct API for their higher level scenarios. Maybe it's premature to push tooling away from NewtonSoft.Json without the higher level parts being implemented. This is not the sort of code I expect to maintain at this level of the stack. I fully appreciate that there are scenarios for dropping down this far, but I don't think the json files read and written by build tooling are a good fit.

[the remark this answers is not in the document]
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: Maybe what I really want are the higher level things like serialization or DOM and not a stream binding for forward-only writing. It's surprising that this doesn't exist, but it's probably still lower than what we'd use if there were a full JSON stack available.

System.Reflection.Metadata (which I maintained) has similar tradeoffs. It is a low level API by design upon which higher levels can be built. I often told people that it was not the direct API for their higher level scenarios. Maybe it's premature to push tooling away from NewtonSoft.Json without the higher level parts being implemented. This is not the sort of code I expect to maintain at this level of the stack. I fully appreciate that there are scenarios for dropping down this far, but I don't think the json files read and written by build tooling are a good fit.

[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: > Maybe what I really want are the higher level things like serialization or DOM and not a stream binding for forward-only writing. It's surprising that this doesn't exist

I am not sure, I follow. It does exist, with plans to build up the feature capabilities:
[JsonSerializer](https://github.com/dotnet/corefx/blob/6c027aebb8fd1c643401bf801083cff5e6d34853/src/System.Text.Json/ref/System.Text.Json.cs#L313-L327)
[JsonDocument](https://github.com/dotnet/corefx/blob/6c027aebb8fd1c643401bf801083cff5e6d34853/src/System.Text.Json/ref/System.Text.Json.cs#L16-L29)

[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: The source package (or at least version 4.6.0-preview3.19128.7 which I pulled from NuGet.org) doesn't contain the `JsonSerializer` class. Looks like the newer versions are all on the blob feed though so we could pull down a newer version with the serialization support.

[the remark this answers is not in the document]
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: Let me rephrase. I wassaying "maybe I really want Serializer or DOM, but I am surprised that forward-only-writer-to-stream is not directly supported".

JsonDocument is read-only today AFAICT from above. Are there plans to be able to build one up programatically and save like XDocument?

(I reach for XDocument a lot more than Xml serialization because I've been burned by serialization many, many times.)

[the remark this answers is not in the document]
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: I did think that there was no JsonSerializer avaialble yet though, happy to see that is in.

[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: > Are there plans to be able to build one up programatically and save like XDocument?

Not for 3.0, but something we are considering adding in the future based on feedback/user scenarios like yours.

[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: FWIW, your concerns and feedback have been heard and hopefully addressed with the recent `Utf8JsonWriter` changes! We ended up making some trade-off between perf and usability to enable the scenarios you mentioned, @nguerrera :)

https://gist.github.com/ahsonkhan/f6f30656717548212693e5eaa49cece5
https://github.com/dotnet/corefx/pull/36961

Presumably, we should re-visit this PR given the new changes!

[the remark this answers is not in the document]
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: Sweet! Thanks!

**Comment by jkoritzinsky** [comment]
@ahsonkhan is there a way to serialize a `System.Collections.Generic.Dictionary<TKey, TValue>` with the JsonSerializer APIs?

**Comment by jkoritzinsky** [comment]
Also @nguerrera @ahsonkhan the newest drop of `Microsoft.Bcl.Json.Sources` uses the `ValueTask<T>` type, which requires a reference to `System.Threading.Tasks.Extensions` on .NET Framework. I know there's been a lot of talk about dependency problems on Framework so I don't want to open a new PR until we've decided what to do.

**Comment by nguerrera** [comment]
Don't open a new PR yet. We need to think about the dependencies more carefully. `That isn't resolved yet. I'm happier with the new API shape, but that is orthogonal to the dependency issues.

**Comment by nguerrera** [comment]
Maybe we should meet up?

**Comment by jkoritzinsky** [comment]
Sounds good to me.

FYI I'm working remotely/oof until Tuesday.

**Comment by nguerrera** [comment]
(And I'm out of office all of next week, and booked all day today.)

**Comment by wli3** [comment]
@jkoritzinsky @nguerrera loop me in

**Comment by ahsonkhan** [comment]
> is there a way to serialize a System.Collections.Generic.Dictionary<TKey, TValue> with the JsonSerializer APIs?

Not currently. It's WIP. See https://github.com/dotnet/corefx/pull/37033 for concrete TKey, TValue. The polymorphic support is still pending.

> Maybe we should meet up?
> FYI I'm working remotely/oof until Tuesday.
> (And I'm out of office all of next week, and booked all day today.)
> loop me in

Sure. Please setup a meeting sometime the week of April 29 or post Build :)
Let's figure it out!

cc @terrajobst
